_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+against different Repository classes (legacy juggler bridge, the new Repository
+implementation we will write in the future) and different connectors (memory,
+MongoDB, MySQL, etc.)
+
+At the moment, the test suite is executed against:
+
+- `memory` connector (as part of `npm test` in `packages/repository-tests`)
+- `mysql` connector - see
+  [`acceptance/repository-mysql`](https://github.com/strongloop/loopback-next/tree/master/acceptance/repository-mysql)
+- `mongodb` connector - see
+  [`acceptance/repository-mongodb`](https://github.com/strongloop/loopback-next/tree/master/acceptance/repository-mongodb)
+
+Please note the shared test suite is very minimal now, we need to beef it up as
+part of follow-up stories.
+
+We also need to enhance our connectors to execute this shared test suite (in
+addition to juggler v3 and v4 tests), i.e. add `@loopback/repository-tests` to
+dev-dependencies and `npm test` of every connector. We should use the same
+approach as we did for juggler v3 and v4, so that in the future, we can run
+tests for multiple `@loopback/repository-tests` and/or `@loopback/repository`
+versions.
+
+Last but not least, let's add Cloudant and PostgreSQL to the connectors tested.
+
+### Edge cases
+
+I have carefully reviewed existing tests in `loopback-datasource-juggler` that
+are related to inclusion of resolved models and discovered few edge to consider.
+
+#### Navigational properties in create/update data
+
+In LB3 test suite, we are testing the following scenario:
+
+```js
+const found = await Category.findOne({include: ['products']});
+found.name = 'new name';
+// important: found.products contains a list of related products
+await found.save();
+// success, found.products was silently ignored
+```
+
+Silently ignoring navigational properties turned out to be problematic in
+practice. Because the framework does not complain, many users are expecting
+related models to be updated as part of the command (e.g. create both `Category`
+and related `Products` in a single call).
+
+For LoopBack 4, we decided to change this behavior and reject such requests with
+an error.
+
+LB3 test suite:
+[loopback-datasource-juggler/test/include.test.js#L104-L141](https://github.com/strongloop/loopback-datasource-juggler/blob/f0a6bd146b7ef2f987fd974ffdb5906cf6a584db/test/include.test.js#L104-L141)
+
+#### Inclusion with custom scope
+
+Besides specifying the relation name to include, it's also possible to specify
+additional `scope` constraints:
+
+- custom `order`, `limit`, `skip` and `fields`
+- additional `scope.where` constraint
+
+For example, the following filter will include only the first active product:
+
+```js
+filter.include = [{
+  relation: 'products',
+  scope: {
+    where: {active: true},
+    limit: 1
+  }
+}
+```
+
+I am proposing to leave this feature out of scope of the initial release.
+However, we should tell the user when they try to use this feature via a 4xx
+error.
```

Comment:
Allowing extra constraint for the `include` is very important and it should be easy to implement as the relational repository already supports the criteria.

Reply:
Agreed, I will be creating a dedicated user story for this feature anyways.

In this spike, I am trying to identify MVP scope. IMO, support for extra constraints is an additive feature that should be implemented after we ship MVP. There will be several features out of MVP scope, we can prioritize and implement them based on their importance, but only after MVP is shipped.

@raymondfeng Do you agree to leave extra constraints out of the MVP scope?